v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/ElasticsearchTransforms.java

```diff
@@ -0,0 +1,326 @@
+   *   <li>{@link WriteToElasticsearchOptions#getUsePartialUpdate()} - use partial updates instead
+   *       of insertions (Default: false).
+   * </ul>
+   *
+   * For {@link ElasticsearchIO#write()} with {@link ValueExtractorFn} if the function returns null
+   * then the index or type provided as {@link WriteToElasticsearchOptions#getIndex()} or {@link
+   * WriteToElasticsearchOptions#getDocumentType()} will be used. For IdFn if function returns null
+   * then the id for the document will be assigned by {@link ElasticsearchIO}.
+   */
+  @AutoValue
+  public abstract static class WriteToElasticsearch extends PTransform<PCollection<String>, PDone> {
+
+    public static Builder newBuilder() {
+      return new AutoValue_ElasticsearchTransforms_WriteToElasticsearch.Builder();
+    }
+
+    public abstract WriteToElasticsearchOptions options();
+
+    @Override
+    public PDone expand(PCollection<String> jsonStrings) {
+
+      ConnectionConfiguration config =
+          ConnectionConfiguration.create(
+              options().getNodeAddresses().split(","),
+              options().getIndex(),
+              options().getDocumentType());
+
+      ElasticsearchIO.Write write =
+          ElasticsearchIO.write()
+              .withConnectionConfiguration(config)
+              .withMaxBatchSize(options().getBatchSize())
+              .withMaxBatchSizeBytes(options().getBatchSizeBytes())
+              .withUsePartialUpdate(options().getUsePartialUpdate());
+
+      if (Optional.ofNullable(options().getMaxRetryAttempts()).isPresent()) {
+        write.withRetryConfiguration(
+            RetryConfiguration.create(
+                options().getMaxRetryAttempts(), getDuration(options().getMaxRetryDuration())));
+      }
+
+      return jsonStrings.apply(
+          "WriteDocuments",
+          write
+              .withIdFn(
+                  ValueExtractorFn.newBuilder()
+                      .setFileSystemPath(options().getIdFnPath())
+                      .setFunctionName(options().getIdFnName())
+                      .build())
+              .withIndexFn(
+                  ValueExtractorFn.newBuilder()
+                      .setFileSystemPath(options().getIndexFnPath())
+                      .setFunctionName(options().getIndexFnName())
+                      .build())
+              .withTypeFn(
+                  ValueExtractorFn.newBuilder()
+                      .setFileSystemPath(options().getTypeFnPath())
+                      .setFunctionName(options().getTypeFnName())
+                      .build()));
+    }
+
+    /** Builder for {@link WriteToElasticsearch}. */
+    @AutoValue.Builder
+    public abstract static class Builder {
+      public abstract Builder setOptions(WriteToElasticsearchOptions options);
+
+      abstract WriteToElasticsearch autoBuild();
+
+      public WriteToElasticsearch build() {
+
+        WriteToElasticsearch write = autoBuild();
```

Comment:
Should we build once all the preconditions are met?

Reply:
Moved autoBuild() to after preconditions are met.